### PR TITLE
[Core] avoid too many cuda context by caching p2p test

### LINKED ISSUE #3821: [Bug]: tp>1 every worker takes memory in every GPU after upgrade to 0.4.0
### Your current environment

```text
$ python collect_env.py
Collecting environment information...
PyTorch version: 2.1.2+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: CentOS Linux 7 (Core) (x86_64)
GCC version: (GCC) 5.4.0
Clang version: Could not collect
CMake version: version 3.27.7
Libc version: glibc-2.17

Python version: 3.11.5 (main, Sep 11 2023, 13:54:46) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-3.10.0-957.21.3.el7.x86_64-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: 11.7.99
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: Tesla V100-SXM2-32GB
GPU 1: Tesla V100-SXM2-32GB
GPU 2: Tesla V100-SXM2-32GB
GPU 3: Tesla V100-SXM2-32GB
GPU 4: Tesla V100-SXM2-32GB
GPU 5: Tesla V100-SXM2-32GB
GPU 6: Tesla V100-SXM2-32GB
GPU 7: Tesla V100-SXM2-32GB

Nvidia driver version: 515.65.01
cuDNN version: /usr/local/cuda-9.2/targets/x86_64-linux/lib/libcudnn.so.7
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                72
On-line CPU(s) list:   0-71
Thread(s) per core:    1
Core(s) per socket:    36
Socket(s):             2
NUMA node(s):          2
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 94
Model name:            Intel(R) Xeon(R) Gold 6133 CPU @ 2.50GHz
Stepping:              3
CPU MHz:               2499.998
BogoMIPS:              4999.99
Hypervisor vendor:     KVM
Virtualization type:   full
L1d cache:             32K
L1i cache:             32K
L2 cache:              4096K
L3 cache:              28160K
NUMA node0 CPU(s):     0-35
NUMA node1 CPU(s):     36-71
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl eagerfpu pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch fsgsbase bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx rdseed adx smap clflushopt xsaveopt xsavec xgetbv1 arat

Versions of relevant libraries:
[pip3] numpy==1.26.3
[pip3] torch==2.1.2+cu118
[pip3] torchaudio==2.1.2+cu118
[pip3] torchvision==0.15.2a0
[pip3] torchviz==0.0.2
[pip3] triton==2.1.0
[conda] blas                      1.0                         mkl    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] cudatoolkit               11.8.0              h4ba93d1_12    conda-forge
[conda] libfaiss                  1.7.4            h2bc3f7f_0_cpu    pytorch
[conda] mkl                       2023.1.0         h213fc3f_46343    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] mkl-service               2.4.0           py311h5eee18b_1    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] mkl_fft                   1.3.8           py311h5eee18b_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] mkl_random                1.2.4           py311hdb19cb5_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] numpy                     1.26.3                   pypi_0    pypi
[conda] torch                     2.1.2+cu118              pypi_0    pypi
[conda] torchaudio                2.1.2+cu118              pypi_0    pypi
[conda] torchvision               0.15.2          cuda118py311h4cc2eb7_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] torchviz                  0.0.2                    pypi_0    pypi
[conda] triton                    2.1.0                    pypi_0    pypiROCM Version: Could not collect
Neuron SDK Version: N/A
vLLM Version: 0.4.0.post1
vLLM Build Flags:
CUDA Archs: Not Set; ROCm: Disabled; Neuron: Disabled
GPU Topology:
GPU0    GPU1    GPU2    GPU3    GPU4    GPU5    GPU6    GPU7    CPU Affinity    NUMA Affinity
GPU0     X      NV1     NV2     NV1     PHB     PHB     PHB     NV2     0-71    0-1
GPU1    NV1      X      NV1     NV2     PHB     PHB     NV2     PHB     0-71    0-1
GPU2    NV2     NV1      X      NV2     PHB     NV1     PHB     PHB     0-71    0-1
GPU3    NV1     NV2     NV2      X      NV1     PHB     PHB     PHB     0-71    0-1
GPU4    PHB     PHB     PHB     NV1      X      NV2     NV2     NV1     0-71    0-1
GPU5    PHB     PHB     NV1     PHB     NV2      X      NV1     NV2     0-71    0-1
GPU6    PHB     NV2     PHB     PHB     NV2     NV1      X      NV1     0-71    0-1
GPU7    NV2     PHB     PHB     PHB     NV1     NV2     NV1      X      0-71    0-1

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks
```

### 🐛 Describe the bug

```
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A    995011      C   python                          27473MiB |
|    0   N/A  N/A    999164      C   ray::RayWorkerVllm                897MiB |
|    0   N/A  N/A    999315      C   ray::RayWorkerVllm                897MiB |
|    0   N/A  N/A    999400      C   ray::RayWorkerVllm                897MiB |
|    1   N/A  N/A    995011      C   python                            897MiB |
|    1   N/A  N/A    999164      C   ray::RayWorkerVllm              27577MiB |
|    1   N/A  N/A    999315      C   ray::RayWorkerVllm                897MiB |
|    1   N/A  N/A    999400      C   ray::RayWorkerVllm                897MiB |
|    2   N/A  N/A    995011      C   python                            897MiB |
|    2   N/A  N/A    999164      C   ray::RayWorkerVllm                897MiB |
|    2   N/A  N/A    999315      C   ray::RayWorkerVllm              27745MiB |
|    2   N/A  N/A    999400      C   ray::RayWorkerVllm                897MiB |
|    3   N/A  N/A    995011      C   python                            897MiB |
|    3   N/A  N/A    999164      C   ray::RayWorkerVllm                897MiB |
|    3   N/A  N/A    999315      C   ray::RayWorkerVllm                897MiB |
|    3   N/A  N/A    999400      C   ray::RayWorkerVllm              27585MiB |
+-----------------------------------------------------------------------------+
```

### PR DESCRIPTION
**Pull request by youkaichao**
It is observed in #3821  that every worker takes memory in every GPU in `_can_p2p`  , because the test will make all process allocate cuda context for every GPU, in total leading to $n * (n-1)$ cuda context.

To avoid this, we can cache the test of p2p test.

# Before this PR (tp=4):

```text
nvidia-smi
Thu Apr 11 15:43:50 2024       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 550.54.15              Driver Version: 550.54.15      CUDA Version: 12.4     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  Tesla V100-SXM2-32GB-LS        On  |   00000000:06:00.0 Off |                    0 |
| N/A   35C    P0             96W /  250W |   29767MiB /  32768MiB |     61%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
|   1  Tesla V100-SXM2-32GB-LS        On  |   00000000:07:00.0 Off |                    0 |
| N/A   38C    P0             96W /  250W |   29609MiB /  32768MiB |     62%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
|   2  Tesla V100-SXM2-32GB-LS        On  |   00000000:0A:00.0 Off |                    0 |
| N/A   37C    P0             78W /  250W |   29717MiB /  32768MiB |     56%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
|   3  Tesla V100-SXM2-32GB-LS        On  |   00000000:0B:00.0 Off |                    0 |
| N/A   35C    P0             75W /  250W |   29657MiB /  32768MiB |     33%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
|   4  Tesla V100-SXM2-32GB-LS        On  |   00000000:85:00.0 Off |                    0 |
| N/A   28C    P0             40W /  250W |       0MiB /  32768MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
|   5  Tesla V100-SXM2-32GB-LS        On  |   00000000:86:00.0 Off |                    0 |
| N/A   29C    P0             41W /  250W |       0MiB /  32768MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
|   6  Tesla V100-SXM2-32GB-LS        On  |   00000000:89:00.0 Off |                    0 |
| N/A   32C    P0             41W /  250W |       0MiB /  32768MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
|   7  Tesla V100-SXM2-32GB-LS        On  |   00000000:8A:00.0 Off |                    0 |
| N/A   31C    P0             47W /  250W |     343MiB /  32768MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
                                                                                         
+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI        PID   Type   Process name                              GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|    0   N/A  N/A   1448586      C   python                                      28810MiB |
|    0   N/A  N/A   1456481      C   ray::RayWorkerVllm                            306MiB |
|    0   N/A  N/A   1458504      C   ray::RayWorkerVllm                            306MiB |
|    0   N/A  N/A   1460718      C   ray::RayWorkerVllm                            306MiB |
|    1   N/A  N/A   1448586      C   python                                        306MiB |
|    1   N/A  N/A   1456481      C   ray::RayWorkerVllm                          28652MiB |
|    1   N/A  N/A   1458504      C   ray::RayWorkerVllm                            306MiB |
|    1   N/A  N/A   1460718      C   ray::RayWorkerVllm                            306MiB |
|    2   N/A  N/A   1448586      C   python                                        306MiB |
|    2   N/A  N/A   1456481      C   ray::RayWorkerVllm                            306MiB |
|    2   N/A  N/A   1458504      C   ray::RayWorkerVllm                          28760MiB |
|    2   N/A  N/A   1460718      C   ray::RayWorkerVllm                            306MiB |
|    3   N/A  N/A   1448586      C   python                                        306MiB |
|    3   N/A  N/A   1456481      C   ray::RayWorkerVllm                            306MiB |
|    3   N/A  N/A   1458504      C   ray::RayWorkerVllm                            306MiB |
|    3   N/A  N/A   1460718      C   ray::RayWorkerVllm                          28700MiB |
+-----------------------------------------------------------------------------------------+
```

GPU blocks: 11762, CPU blocks: 2048
Throughput: 8.17 requests/s, 3932.45 tokens/s

# After this PR (tp=4):

```text
✗ nvidia-smi
Thu Apr 11 15:52:19 2024       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 550.54.15              Driver Version: 550.54.15      CUDA Version: 12.4     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  Tesla V100-SXM2-32GB-LS        On  |   00000000:06:00.0 Off |                    0 |
| N/A   36C    P0             87W /  250W |   29717MiB /  32768MiB |     72%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
|   1  Tesla V100-SXM2-32GB-LS        On  |   00000000:07:00.0 Off |                    0 |
| N/A   38C    P0             87W /  250W |   29559MiB /  32768MiB |     66%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
|   2  Tesla V100-SXM2-32GB-LS        On  |   00000000:0A:00.0 Off |                    0 |
| N/A   37C    P0             85W /  250W |   29667MiB /  32768MiB |     61%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
|   3  Tesla V100-SXM2-32GB-LS        On  |   00000000:0B:00.0 Off |                    0 |
| N/A   35C    P0             79W /  250W |   29607MiB /  32768MiB |     64%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
|   4  Tesla V100-SXM2-32GB-LS        On  |   00000000:85:00.0 Off |                    0 |
| N/A   28C    P0             40W /  250W |       0MiB /  32768MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
|   5  Tesla V100-SXM2-32GB-LS        On  |   00000000:86:00.0 Off |                    0 |
| N/A   29C    P0             41W /  250W |       0MiB /  32768MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
|   6  Tesla V100-SXM2-32GB-LS        On  |   00000000:89:00.0 Off |                    0 |
| N/A   30C    P0             41W /  250W |       0MiB /  32768MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
|   7  Tesla V100-SXM2-32GB-LS        On  |   00000000:8A:00.0 Off |                    0 |
| N/A   29C    P0             42W /  250W |       0MiB /  32768MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
                                                                                         
+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI        PID   Type   Process name                              GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|    0   N/A  N/A   1491851      C   python                                      29706MiB |
|    1   N/A  N/A   1496915      C   ray::RayWorkerVllm                          29548MiB |
|    2   N/A  N/A   1497165      C   ray::RayWorkerVllm                          29656MiB |
|    3   N/A  N/A   1497328      C   ray::RayWorkerVllm                          29596MiB |
+-----------------------------------------------------------------------------------------+
```
GPU blocks: 12222, CPU blocks: 2048
Throughput: 8.21 requests/s, 3953.50 tokens/s

# Conclusion

Fight really hard to save 306MiB * tp * tp memory.

### CODE REVIEW

File: vllm/distributed/utils.py [hunk 1]
```diff
@@ -46,3 +55,60 @@ def split_tensor_along_last_dim(
...
+def _can_actually_p2p(idx_a, idx_b):
+    dev_i = f"cuda:{idx_a}"
+    dev_j = f"cuda:{idx_b}"
+    a = torch.randn(5, device=dev_i) + 123.0
+    b = a.to(dev_j)
+    c = b.to(dev_i)
+    return torch.all(a == c).cpu().item()
+
+
+_gpu_p2p_access_cache = None
```
Comment (rkooo567) [inline vllm/distributed/utils.py:72]: add type?

File: vllm/distributed/utils.py [hunk 2]
```diff
@@ -46,3 +55,60 @@ def split_tensor_along_last_dim(
...
+    if _gpu_p2p_access_cache is not None:
+        return _gpu_p2p_access_cache[f"{i}->{j}"]
+
+    is_distributed = dist.is_initialized()
+
+    num_dev = torch.cuda.device_count()
+    cuda_visible_devices = os.environ.get("CUDA_VISIBLE_DEVICES", None)
+    if cuda_visible_devices is None:
+        cuda_visible_devices = ",".join(str(i) for i in range(num_dev))
+    path = os.path.expanduser(
```
Comment (rkooo567) [inline vllm/distributed/utils.py:90]: nit; consider using pathlib instead (https://docs.python.org/3/library/pathlib.html)?

File: vllm/distributed/utils.py [hunk 3]
```diff
@@ -46,3 +55,60 @@ def split_tensor_along_last_dim(
...
+        f"~/.config/vllm/gpu_p2p_access_cache_for_{cuda_visible_devices}.json")
+    os.makedirs(os.path.dirname(path), exist_ok=True)
+    if (not is_distributed or dist.get_rank()== 0) \
```
Comment (rkooo567) [inline vllm/distributed/utils.py:93]: nit: I believe `\` is not necessary here

**Review by rkooo567 — COMMENTED** [review]
QQ: why do we write the cache to a file? Is it so that all the workers can access the cache info? 

In this case, what's happening if there are 2 vllm instances running at the same time? (isn't the cache file overwriting each other, or is it just to check gpu to gpu accessibility, so that they can share the same result?)

File: vllm/distributed/utils.py (hunk 3 above)
Comment (rkooo567) [inline vllm/distributed/utils.py:93]: probably the best adding a file lock here so that we can avoid 2 instances overwriting the file?

Comment (youkaichao) [reply vllm/distributed/utils.py:93]: for two instances of vllm, they will have different file path because of different `cuda_visible_devices`.

Comment (youkaichao) [reply vllm/distributed/utils.py:93]: we have to use `\` because we have a line limit of 79 char I think :(

**Comment by youkaichao** [comment]
> why do we write the cache to a file? Is it so that all the workers can access the cache info?

If we don't cache it into a file, every master process still needs to initialize cuda context in every gpu, that will lead to `tp * 306MiB` memory cost. In one machine, the p2p access for a fixed set of GPUs should not change. So we can safely cache it into a file, and later even master process does not need to pay the cost.

> what's happening if there are 2 vllm instances running at the same time?

Note that the cache file name is suffixed by `cuda_visible_devices`. So 2 vllm instances will not have conflict.

File: vllm/distributed/utils.py (hunk 1 above)
Comment (youkaichao) [reply vllm/distributed/utils.py]: fixed in [72e1128](https://github.com/vllm-project/vllm/pull/4021/commits/72e1128965ea30f7d48256ab8ccb3b4547f60de8).

File: vllm/distributed/utils.py (hunk 2 above)
Comment (youkaichao) [reply vllm/distributed/utils.py:90]: I think this is fine, we have similar code at:

https://github.com/vllm-project/vllm/blob/e46a60aa4c90cf3dfd9b90782f2eeabbda935eef/vllm/distributed/device_communicators/pynccl.py#L42

**Comment by WoosukKwon** [comment]
QQ: Can we directly use [cudaIpcOpenMemHandle](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DEVICE.html#group__CUDART__DEVICE_1g01050a29fefde385b1042081ada4cde9) instead?

**Comment by youkaichao** [comment]
> QQ: Can we directly use [cudaIpcOpenMemHandle](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DEVICE.html#group__CUDART__DEVICE_1g01050a29fefde385b1042081ada4cde9) instead?

We should seek help from @hanzhi713 , I'm not familiar with this :(

**Comment by cadedaniel** [comment]
> > QQ: Can we directly use [cudaIpcOpenMemHandle](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DEVICE.html#group__CUDART__DEVICE_1g01050a29fefde385b1042081ada4cde9) instead?
> 
> We should seek help from @hanzhi713 , I'm not familiar with this :(

We can use `cudaDeviceCanAccessPeer`, accessible with cupy: https://docs.cupy.dev/en/stable/reference/generated/cupy.cuda.runtime.deviceCanAccessPeer.html

**Comment by rkooo567** [comment]
Makes sense! Maybe it'd be great to have a simple comment in the function for the motivation of writing it to a file!

**Comment by youkaichao** [comment]
Technically many functions can be used to detect p2p access, but the details of which function will cost additional cuda context is extemely vague. [cudaIpcOpenMemHandle](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DEVICE.html#group__CUDART__DEVICE_1g01050a29fefde385b1042081ada4cde9) or [cudaDeviceCanAccessPeer](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__PEER.html#group__CUDART__PEER_1g4db0d04e44995d5c1c34be4ecc863f22) might work, but it is hard to say if they will place `tp * tp` cuda context.

That said, I think caching the result is a universal way, regardless of how we detect p2p access.

**Comment by youkaichao** [comment]
Anyone stamp my PR? Or any additional modification required?

**Comment by esmeetu** [comment]
1. multi-node seems not working?
2. Should we delete this cache after p2p test done?

**Review by rkooo567 — APPROVED** [review]

**Comment by rkooo567** [comment]
Multi node is not working with vllm anyway, so not sure if we should handle it in this PR

**Comment by youkaichao** [comment]
> Multi node is not working with vllm anyway

This is not correct. vllm indeed supports multi-node.

@esmeetu I added the case for multi-node, by letting one process per node to create the cache (i.e. `local_rank == 0`). Could you please test if this works for multi-node setting?

**Comment by hanzhi713** [comment]
Technically either cudaDeviceCanAccessPeer or cudaIpcOpenMemHandle would suffice. cudaDeviceCanAccessPeer is called by torch.cuda.can_device_access_peer. However, cuda driver sometimes is buggy and might report p2p being supported even though it's not. This can occur on 3090 and 4090. Thus, we need to perform actual p2p copies and check whether the result is correct to mitigate the driver bug.

**Comment by youkaichao** [comment]
I can confirm `cudaDeviceCanAccessPeer ` will cost about `tp * tp * 300MB` memory. Not sure if `cudaIpcOpenMemHandle` behaves the same.

Either way, the caching in this PR is reasonable. p2p access pattern between GPUs seldom change.

**Comment by esmeetu** [comment]
> > Multi node is not working with vllm anyway
> 
> This is not correct. vllm indeed supports multi-node.
> 
> @esmeetu I added the case for multi-node, by letting one process per node to create the cache (i.e. `local_rank == 0`). Could you please test if this works for multi-node setting?

Multi-node was not supported when using custom-all-reduce. So this PR looks good to me.

**Review by esmeetu — APPROVED** [review]

**Pull request merged**